### PR TITLE
feat: Support the compiler directive statements

### PR DESCRIPTION
**Pull request by ap891843**
This PR addresses the proper error message for the following compiler directives:
[IBM compiler directive](https://www.ibm.com/support/knowledgecenter/SS6SG3_6.3.0/lr/ref/rlcdsuse.html)
- *CONTROL(*CBL)
- EJECT
- PROCESS
- SKIP1/2/3
- TITLE
- USE
- ENTER

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

### CODE REVIEW

File: server/src/test/java/com/broadcom/lsp/cobol/usecases/engine/UseCasePreprocessorListener.java [hunk 1]
```diff
@@ -477,7 +477,9 @@ private void registerDiagnostics(Range range, List<DiagnosticContext> diagnostic
             .get(documentUri)
             .add(
                 new Diagnostic(
-                    range, it.getMessage(), it.getSeverity(), it.getSource(), it.getCode()));
+                        // honour the range provided by tester.
```
Comment (ap891843) [inline server/src/test/java/com/broadcom/lsp/cobol/usecases/engine/UseCasePreprocessorListener.java:480]: Controversial change. I will work on this if the team feels this is not the right approach.
Basically, had to add this as I was unable to add diagnostics across a set of identifiers( which are clubbed together as a single token rule) .

File: server/src/main/java/com/broadcom/lsp/cobol/core/engine/CobolLanguageEngine.java [hunk 2]
```diff
@@ -157,7 +160,42 @@ CobolParser getCobolParser(CommonTokenStream tokens) {
 
   @NonNull
   private Function<SyntaxError, SyntaxError> convertError(@NonNull Map<Token, Locality> mapping) {
-    return err -> err.toBuilder().locality(mapping.get(err.getOffendedToken())).build();
+    return err ->
+        err.toBuilder()
+            .locality(
+                mapping.getOrDefault(
+                    err.getOffendedToken(), getNearestLocality(err.getOffendedToken(), mapping)))
```
Comment (ap891843) [inline server/src/main/java/com/broadcom/lsp/cobol/core/engine/CobolLanguageEngine.java:167]: Another controversial change. 
Reason: In case we want to highlight an error from a lexer action based on tokens, which is a subset of a single token.
For e.g. suppose we have a single token, SINGLETOKEN, and based on lexer action, it is decided that there is an error on the "LE" part of that token. In this case, we want to highlight the entire token with an error message generated by the lexer action.

File: server/src/main/resources/resourceBundles/messages_en.properties [hunk 3]
```diff
@@ -40,3 +40,8 @@ db2SqlParser.size=%s not allowed. Size should be in GB.
 db2SqlParser.maxIntValue=%s not allowed. Allowed range is 0<k<=32767.
 db2SqlParser.currectQueryAccelaration=%s not valid. Must be of format %d%d%d%d.%d
 cobolParser.subSchemaNameLength=The length %s is not allowed. Allowed values are 16, 18.
+ContinuationLineTransformation.compilerDirectiveContinued=Compiler directives cannot be continued on another line.
```
Comment (ap891843) [inline server/src/main/resources/resourceBundles/messages_en.properties:43]: @zimlu02  Please check these error messages.

File: server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4 [hunk 4]
```diff
@@ -15,9 +15,56 @@
...
+    reportWrongArguments("lexer.langMissingEnterDirective");
+    }
+    }
+}
 
+// compiler directive tokens
+TITLESTATEMENT : (T I T L E ' '+ .*? NEWLINE) {isTitlePresent();} -> channel(TECHNICAL);
+CONTROL_DIRECTIVE: ASTERISKCHAR (CONTROL | C B L) ((' '| COMMACHAR) (SOURCE | NO SOURCE | LIST | NO LIST | MAP | NO
+MAP | IDENTIFIER? {reportWrongArguments("lexer.controlDirectiveWrongArgs");}))+ DOT?-> channel
+(TECHNICAL);
```
Comment (grianbrcom) [inline server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4:66]: Please, use indentation here.

File: server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolPreprocessorLexer.g4 [hunk 5]
```diff
@@ -22,6 +22,11 @@ lexer grammar CobolPreprocessorLexer;
         lastTokenText = token.getText();
     }
 }
+// compiler directive tokens
+TITLESTATEMENT : (T I T L E ' '+ .*? NEWLINE)  ;
+CONTROL_DIRECTIVE: ASTERISKCHAR (C O N T R O L | C B L) ((' '| COMMACHAR) (SOURCE | NO SOURCE | LIST | NO LIST | MAP
+| NO MAP | IDENTIFIER? ))+  DOT?;
```
Comment (grianbrcom) [inline server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolPreprocessorLexer.g4:28]: Indent it, please.

File: server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4 [hunk 6]
```diff
@@ -15,9 +15,59 @@
 lexer grammar CobolLexer;
 import CICSLexer;
 channels{TECHNICAL}
+@header {
+  import java.util.regex.Matcher;
+  import java.util.regex.Pattern;
+  import org.apache.commons.lang3.StringUtils;
+}
+@lexer::members {
+private Pattern titlePattern = Pattern.compile("TITLE\\s*(.*?)\\.?\\s\\n.*", Pattern.CASE_INSENSITIVE);
```
Comment (asatklichov) [inline server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4:24]: Making 'titlePattern' static final, will be more performant (pattern compiled only once).

File: server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4 [hunk 7]
```diff
@@ -15,9 +15,59 @@
...
+                if(matchedTitle.isEmpty())
+                reportWrongArguments("lexer.titleCompilerDirective");
+                }
+          }
+        }
+    private void reportWrongArguments(String msg) {
+        ANTLRErrorListener listener = getErrorListenerDispatch();
+        String text = _input.getText(Interval.of(_tokenStartCharIndex, _input.index()));
+	    int stop = _input.index();
+	    if(text.contains("\r")) {
```
Comment (asatklichov) [inline server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4:40]: What about \n, \r\n?

File: server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4 [hunk 8]
```diff
@@ -15,9 +15,59 @@
...
+    if(StringUtils.isBlank(input.substring(5, input.length()-1))) {
+    reportWrongArguments("lexer.langMissingEnterDirective");
+    }
+    }
+}
 
+// compiler directive tokens
+TITLESTATEMENT : (T I T L E ' '+ .*? NEWLINE) {isTitlePresent();} -> channel(TECHNICAL);
+
+CONTROL_DIRECTIVE: ASTERISKCHAR (CONTROL | C B L) ((' '| COMMACHAR)
```
Comment (asatklichov) [inline server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4:65]: Any specific reason not defining this (CONTROL_DIRECTIVE) as rule in parser? (want to use benefit of channel ?)

File: server/src/main/resources/resourceBundles/messages_en.properties (hunk 3 above)
Comment (zimlu02) [reply server/src/main/resources/resourceBundles/messages_en.properties:43]: The first 2 miss "is", third has "is". Other from that looks OK. I would rather @zeibura to check this, pls.

Comment (zeibura) [reply server/src/main/resources/resourceBundles/messages_en.properties:43]: What do these error messages actually look like? Can you send me a screenshot of one of them?

Comment (ap891843) [reply server/src/main/resources/resourceBundles/messages_en.properties:43]: image
image

Comment (zeibura) [reply server/src/main/resources/resourceBundles/messages_en.properties:43]: I would suggest we do them like this

db2SqlParser.validValueMsg=Syntax error on %s expected %s
db2SqlParser.maxLength=Max length limit of %s bytes allowed for %s.
db2SqlParser.piecesize=Invalid size %s. Specify a size in KB, MB or GB.
db2SqlParser.size=Invalid size %s. Specify a size in GB.
db2SqlParser.maxIntValue=Invalid value %s. Allowed range is 0<k<=32767.
db2SqlParser.currectQueryAccelaration=Invalid format %s. Must be in format %d%d%d%d.%d

Comment (zeibura) [reply server/src/main/resources/resourceBundles/messages_en.properties:43]: cobolParser.subSchemaNameLength=Invalid length %s. Allowed values are 16, 18.
ContinuationLineTransformation.compilerDirectiveContinued=Compiler directives cannot be continued on another line. 
CompilerDirectivesTransformation.sequenceNumber=The first character of the sequence number must be numeric.
lexer.controlDirectiveWrongArgs=Invalid argument for CONTROL compiler directives.
lexer.langMissingEnterDirective=LanguageName missing for ENTER compiler directive.
lexer.titleCompilerDirective=Title statement missing for TITLE compiler directive.

File: server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4 (hunk 7 above)
Comment (ap891843) [reply server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4]: That won't be a problem, we just need a subset of token.

File: server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4 (hunk 8 above)
Comment (ap891843) [reply server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4]: Yes, these directives can come anywhere in a source code. something like whitespace. So, we want to throw them in a diff channel/skip. i.e. we don't want to process these tokens.

File: server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4 (hunk 7 above)
Comment (ap891843) [reply server/src/main/antlr4/com/broadcom/lsp/cobol/core/parser/CobolLexer.g4]: The initial inclination was to get the exact token and have our own custom tokens (extends CommonToken).  But changed it coz it was difficult to get the token.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CompilerDirectivesTransformation.java [hunk 9]
```diff
@@ -40,10 +52,36 @@ public String apply(@NonNull String line) {
...
+
   private String cutTooLongString(String line) {
     return line.length() == 80 ? line.substring(0, 72).trim() : line;
   }
 
+  private boolean isSequenceNumberFormatCorrect(String line) {
+    int lineContentStart = getLineContentStart(line);
+    // issue error the sequence must start with a number.
+    return lineContentStart < CobolLinesTransformation.INDICATOR_AREA
+        || !StringUtils.isNotBlank(line.substring(0, lineContentStart))
```
Comment (asatklichov) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CompilerDirectivesTransformation.java:81]: There is isBlank blank method instead of !StringUtils.isNotBlank

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessor.g4 [hunk 10]
```diff
@@ -207,6 +209,18 @@ familyPhrase
    ;
 
 replaceable
+   : literal | cobolWord | charDataLine | functionCall
+   ;
+
+functionCall
```
Comment (temanbrcom) [inline server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessor.g4:215]: Why did you add it here?

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessorLexer.g4 [hunk 11]
```diff
@@ -22,6 +22,14 @@ lexer grammar CobolPreprocessorLexer;
         lastTokenText = token.getText();
     }
 }
+// compiler directive tokens
+TITLESTATEMENT : (T I T L E ' '+ .*? NEWLINE)  ;
+
+CONTROL_DIRECTIVE: ASTERISKCHAR (C O N T R O L | C B L) ((' '| COMMACHAR)
```
Comment (temanbrcom) [inline server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessorLexer.g4:28]: I see the list of directives is incomplete, is it intentional?

File: server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java [hunk 12]
```diff
@@ -157,7 +157,42 @@ CobolParser getCobolParser(CommonTokenStream tokens) {
...
+                        && entry.getKey().getLine() == referenceToken.getLine()
+                        && entry.getKey().getStartIndex() <= referenceToken.getStartIndex()
+                        && entry.getKey().getStopIndex() >= referenceToken.getStopIndex())
+            .map(Map.Entry::getValue)
+            .findFirst();
+
+    return nearestToken.orElse(null); // null is safe here.
+  }
+
+  private String convertErrorMessage(String suggestion) {
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java:188]: I would rather catch the exceptions on the service side

File: server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java [hunk 13]
```diff
@@ -157,7 +157,42 @@ CobolParser getCobolParser(CommonTokenStream tokens) {
...
+                        && entry.getKey().getCharPositionInLine()
+                            == referenceToken.getCharPositionInLine()
+                        && entry.getKey().getLine() == referenceToken.getLine()
+                        && entry.getKey().getStartIndex() <= referenceToken.getStartIndex()
+                        && entry.getKey().getStopIndex() >= referenceToken.getStopIndex())
+            .map(Map.Entry::getValue)
+            .findFirst();
+
+    return nearestToken.orElse(null); // null is safe here.
+  }
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java:186]: Logically, it might be a part of `LocalityMappingUtils`

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessorLexer.g4 (hunk 11 above)
Comment (ap891843) [reply server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessorLexer.g4:28]: Only mentioned the tokens which were a combination of multiple identifiers. Mentioned these avoid mismatch in preprocessor and cobol lexer.

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolLexer.g4 [hunk 14]
```diff
@@ -15,9 +15,59 @@
 lexer grammar CobolLexer;
 import CICSLexer;
 channels{TECHNICAL}
+@header {
+  import java.util.regex.Matcher;
+  import java.util.regex.Pattern;
+  import org.apache.commons.lang3.StringUtils;
+}
+@lexer::members {
+private Pattern titlePattern = Pattern.compile("TITLE\\s*(.*?)\\.?\\s\\n.*", Pattern.CASE_INSENSITIVE);
```
[the remark this answers is not in the document]
Comment (ap891843) [reply server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolLexer.g4]: done.

File: server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java (hunk 12 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java]: Make sense. Done.

File: server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java (hunk 13 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java]: Done

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolLexer.g4 (hunk 14 above)
[the remark this answers is not in the document]
Comment (ap891843) [reply server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolLexer.g4]: Yes, modified a use-case as well to test this.

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessor.g4 (hunk 10 above)
Comment (ap891843) [reply server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessor.g4:215]: It's part of the syntax. We don't want to support this functionality but just the syntax (as discussed with Scott).
I am not sure it works. I will keep this conversation open, till I have a use-case for it or it works (syntactically)

image

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderDelegate.java [hunk 15]
```diff
@@ -15,19 +15,34 @@
...
-  @NonNull
-  String apply(@NonNull String line);
+    /**
+     * Convert the line and apply any transformations if needed.
+     *
+     * @param line the initial line
+     * @return converting result
+     */
+    @NonNull
+    String apply(@NonNull String line);
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderDelegate.java:35]: Seems this method is unused

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderImpl.java [hunk 16]
```diff
@@ -162,7 +173,7 @@ private SyntaxError createError(
                     .uri(uri)
                     .range(
                         new Range(new Position(lineNumber, start), new Position(lineNumber, stop)))
-                    .recognizer(getClass())
+                    .recognizer(MethodHandles.lookup().lookupClass())
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderImpl.java:176]: Why?

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/CobolLinesTransformation.java [hunk 17]
```diff
@@ -25,6 +25,9 @@
  */
 public interface CobolLinesTransformation {
 
+  int START_INDEX_AREA_A = 4;
+  int END_INDEX_CONTENT_AREA_A = 10;
+  int INDICATOR_AREA = 7;
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/CobolLinesTransformation.java:30]: You may move them to the `ProcessingConstants` class

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java [hunk 18]
```diff
@@ -85,12 +86,73 @@ private SyntaxError checkContinuationLine(String uri, int lineNumber, CobolLine
...
+        return compilerDirectiveContinuedError;
+      }
+
       // invoke method for noContentInAreaA
       return checkContentAreaAWithContinuationLine(cobolLine, uri, lineNumber);
     }
     return null;
   }
 
+  private SyntaxError isCompilerDirectiveContinued(
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java:102]: Should start with "check"

File: server/src/test/java/org/eclipse/lsp/cobol/usecases/TestControlCompilerDirective.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.junit.jupiter.api.Test;
+
+/** This class test the *CONTROL (*CBL) compiler directives. */
+class TestControlCompilerDirective {
```
[the remark this answers is not in the document]
Comment (ap891843) [reply server/src/test/java/org/eclipse/lsp/cobol/usecases/TestControlCompilerDirective.java:29]: PROCESS(CBL) should be placed before ID division [PROCESS_LINK](https://www.ibm.com/support/knowledgecenter/SS6SG3_6.1.0/lr/ref/rlcdscbl.html) 

*CONTROL/*CBL [CBL_LINK](https://www.ibm.com/support/knowledgecenter/SS6SG3_6.3.0/lr/ref/rlcdscon.html)
I am still working on fixing *CBL between process and end as part of another story.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderImpl.java (hunk 16 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderImpl.java:176]: Don't remember this change.  Definitely not required, will get rid of this.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CompilerDirectives.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+package org.eclipse.lsp.cobol.core.preprocessor.delegates.reader;
+
+/**
+ * Enum for supported Compiler directives. These compiler directives can't be continued to the next
+ * line.
+ */
+public enum CompilerDirectives {
+  TITLE("TITLE"),
```
[the remark this answers is not in the document]
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CompilerDirectives.java]: Done

File: server/src/test/java/org/eclipse/lsp/cobol/usecases/TestReplacingNotAppliedRecursively.java [hunk 21]
```diff
@@ -33,19 +34,27 @@
...
   private static final String STRUCT1_NAME = "STRUCT1";
 
+  /**
+   * Ref:https://www.ibm.com/support/knowledgecenter/en/SS6SG3_6.3.0/lr/ref/rlcdscop.html#rlcdscop__CDSREP
+   *
+   * When the REPLACING phrase is specified for a COPY statement that appears in a chain of nested
+   * COPY statements, the REPLACING phrase applies to all library text that is included by COPY
+   * statements nested under the COPY statement that has the REPLACING phrase.
+   */
+  @Disabled("not compliant with z/OS 6.3.0 docs")
```
[the remark this answers is not in the document]
Comment (ap891843) [reply server/src/test/java/org/eclipse/lsp/cobol/usecases/TestReplacingNotAppliedRecursively.java]: deleted.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/CobolLinesTransformation.java (hunk 17 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/CobolLinesTransformation.java]: done.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java (hunk 18 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java]: done.

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolLexer.g4 [hunk 22]
```diff
@@ -15,9 +15,59 @@
...
+		   stop--;
+	    }
+        CommonToken lspToken = new CommonTokenFactory().create(_tokenFactorySourcePair, _type, text,
+	         					_channel, _input.index()-text.length()+1,
+	         					stop,
+	        					_tokenStartLine, _tokenStartCharPositionInLine);
+        lspToken.setTokenIndex(_tokenStartCharPositionInLine);
+        lspToken.setText(text.replace("\r","").replace("\n",""));
+        listener.syntaxError(this, lspToken, _tokenStartLine,
+         	         _tokenStartCharPositionInLine, msg, null);
```
[the remark this answers is not in the document]
Comment (ap891843) [reply server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolLexer.g4:50]: Yes, make sense to simplify this. I will raise a separate PR this at last while finalizing compiler directives.

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessor.g4 (hunk 10 above)
Comment (ap891843) [reply server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessor.g4:215]: covered with a use-case test.

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessorLexer.g4 (hunk 11 above)
Comment (temanbrcom) [reply server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessorLexer.g4:28]: Also, why do you write `C B L` this way when it is declared explicitly? I would recommend declaring also the `CONTROL`

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java [hunk 23]
```diff
@@ -102,12 +178,29 @@ private String getReplacementPattern(@NonNull String text) {
...
   @NonNull
-  private String extractPseudoText(@NonNull String text) {
-    return text.trim().replaceAll("^==", "").replaceAll("==$", "").replaceAll(" +", " ").trim();
+  private String extractPseudoText(@NonNull String text, @NonNull boolean isOperandOne) {
+    String processedText =
+        text.trim().replaceAll("^==", "").replaceAll("==$", "").replaceAll(" +", " ");
+    if (isOperandOne && (StringUtils.countMatches(processedText, ",") == 1))
+      return processedText.trim();
+    if (isOperandOne && (StringUtils.countMatches(processedText, ";") == 1))
+      return processedText.trim();
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java:197]: May they be combined?

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderImpl.java [hunk 24]
```diff
@@ -150,8 +151,17 @@ public CobolLineReaderImpl(CobolLineReaderDelegate delegate, MessageService mess
...
+   * @param uri the document URI
+   * @param message the error message
+   * @param lineNumber the error lineNumber
+   * @param start the error start position
+   * @param stop the error stop position
+   * @return the {@link SyntaxError}
+   */
   @NonNull
-  private SyntaxError createError(
+  public static SyntaxError createError(
```
Comment (temanbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderImpl.java:164]: Why is it public?

Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/reader/CobolLineReaderImpl.java]: I agree. It was me being lazy. Corrected.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java (hunk 23 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java]: done

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java [hunk 25]
```diff
@@ -85,12 +86,73 @@ private SyntaxError checkContinuationLine(String uri, int lineNumber, CobolLine
...
+  private boolean isCompilerDirectiveStatement(CobolLine line) {
+    if (Objects.isNull(line.getPredecessor())) return false;
+    String predecessorContentArea = line.getPredecessor().getContentArea();
+    return Stream.of(CompilerDirectives.values())
+            .map(CompilerDirectives::getDirective)
+            .anyMatch(predecessorContentArea.split("\\s+")[0].toUpperCase()::equals)
+        || isPseudoDelimiterContinued(line, predecessorContentArea);
+  }
+
+  private boolean isPseudoDelimiterContinued(CobolLine line, String predecessorContentArea) {
```
[the remark this answers is not in the document]
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java:133]: modified. Please check

File: server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessorLexer.g4 (hunk 11 above)
Comment (ap891843) [reply server/src/main/antlr4/org/eclipse/lsp/cobol/core/parser/CobolPreprocessorLexer.g4]: done.

**Review by temanbrcom — APPROVED** [review]
🍊

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java [hunk 26]
```diff
@@ -85,12 +86,85 @@ private SyntaxError checkContinuationLine(String uri, int lineNumber, CobolLine
...
+    return lastValidChar;
+  }
+
+  private boolean isQuotesBalanced(String predecessorContentArea) {
+    int quotesCount = StringUtils.countMatches(predecessorContentArea, "\"");
+    int singleQuotesCount = StringUtils.countMatches(predecessorContentArea, "'");
+
+    return isEven(quotesCount)
+        || isEven(singleQuotesCount)
+        || (quotesCount + singleQuotesCount == 0);
```
Comment (grianbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java:160]: Why `||`? Do you need all types of quotes to be balanced or just one?

BTW, you can replace `isEven(quotesCount)` by `quotesCount % 2 == 0`.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/LocalityMappingUtils.java [hunk 27]
```diff
@@ -184,4 +185,35 @@ private boolean tokenMatches(String token, String position) {
...
+   * @return Locality for a passed token.
+   */
+  public Optional<Locality> getNearestLocality(
+      Token referenceToken, @NonNull Map<Token, Locality> mapping) {
+    if (Objects.isNull(referenceToken)) return Optional.empty();
+    return mapping.entrySet().stream()
+        .filter(
+            entry ->
+                StringUtils.normalizeSpace(entry.getKey().getText())
+                        .contains(StringUtils.normalizeSpace(referenceToken.getText()))
```
Comment (grianbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/LocalityMappingUtils.java:210]: You can extract it out of the filter. Thus you normalizeSpace for token only once, you don't need to do it for each mapping entry.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java [hunk 28]
```diff
@@ -120,4 +209,30 @@ private String replace(@NonNull String text, @NonNull Pair<String, String> patte
...
+  }
+
+  @NonNull
+  private ProcessedSearchClause getProcessedSearchClause(@NonNull String clause) {
+    SearchPattern searchPattern = SearchPattern.EXACT;
+    Matcher matcher = LEAD_OR_TRAIL_CLAUSE.matcher(clause);
+    if (matcher.matches()) {
+      if (matcher.group(1).trim().equalsIgnoreCase("LEADING"))
+        searchPattern = SearchPattern.STARTS_WITH;
+      else if (matcher.group(1).trim().equalsIgnoreCase("TRAILING"))
```
Comment (grianbrcom) [inline server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java:225]: You don't need to trim the group if you change your pattern to `"\\s*(LEADING|TRAILING).*"` (without space after | sign).

**Review by grianbrcom — APPROVED** [review]
:compass:

**Review by asatklichov — APPROVED** [review]
🏯

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java (hunk 26 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java:160]: 👍  isEven was stupid 😄 
I am re-thinking on this. Need to check the below statements
```
"Hello, I don't know"

' "Hello" , I don't know ' 
```

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java (hunk 28 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/ReplacingServiceImpl.java:225]: done

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/LocalityMappingUtils.java (hunk 27 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/util/LocalityMappingUtils.java]: Done.

File: server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java (hunk 26 above)
Comment (ap891843) [reply server/src/main/java/org/eclipse/lsp/cobol/core/preprocessor/delegates/transformer/ContinuationLineTransformation.java]: Done. Remove that logic completely.
ref: https://www.ibm.com/support/knowledgecenter/SS6SG3_6.1.0/lr/ref/rlfmtcon.html
Topic: Continuation of alphanumeric and national literals

**Review by grianbrcom — APPROVED** [review]
:gear:
